File: 4-tic-tac-boom/solution_comments.ml
(*
  # The Scripting Language Tournament by InfiniteCoder01
  Challenge 4 - Tic Tac Boom
  Date: 28th August 2025

  Entry:  OCaml
  Author: Nigel Withrow <nigelwithrow78@gmail.com>

  Instructions:
  + Have system-wide Rust & Cargo installation
  + `$ cd 4-tic-tac-boom`
  + `$ nix develop`
  + `$ just tic-tac-boom`
  + `$ ./target/release/tic-tac-boom ./solution.ml`
*)

module Game = struct
  (*
    Inductive definition of non-zero unsigned integer
    Used for better pattern matching on X and O's remaining lifespan
  *)
  type nzuint = One | S of nzuint

  (* Every spawned X or O lives for 4 rounds (game tick happens right after
     spawning) *)
  let four = S (S (S One))

  (* Empty or O or X *)
  type cell = E | O of nzuint | X of nzuint

  let string_of_cell = function E -> ' ' | O _ -> 'o' | X _ -> 'x'

  (* Game board *)
  type board = cell Array.t

  let string_of_board board =
    board |> Array.to_seq |> Seq.map string_of_cell |> String.of_seq

  (* Board and its string representation used for diff'ing with input *)
  type t = { board : board; mutable str : string }

  let empty =
    { board = Seq.repeat E |> Seq.take 9 |> Array.of_seq; str = "         " }

  (*
    Tick lifespan of Xs and Os
  *)
  let tick_x self : unit =
    Array.iteri
      (fun i cell ->
        self.board.(i) <-
          (match cell with E -> E | X One -> E | X (S n) -> X n | O n -> O n))
      self.board

  let tick_o self =
    Array.iteri
      (fun i cell ->
        self.board.(i) <-
          (match cell with E -> E | O One -> E | O (S n) -> O n | X n -> X n))
      self.board

  (*
    Add moves from X and O
  *)
  let move_x i self =
    (match self.board.(i) with
    | E -> ()
    | _ -> raise @@ Failure "X trying to play at occupied space");
    self.board.(i) <- X four

  let move_o i self =
    (match self.board.(i) with
    | E -> ()
    | _ -> raise @@ Failure "O trying to play at occupied space");
    self.board.(i) <- O four
end

(*
  Strategies decided what move we should play
*)
module Strategy = struct
  open Game

  (*
    Match a provided 3-cell permutation to see if any of its combinations occurs
    on the game board, returning the indexes of the matched cells if matched
  *)
  let match3 board (a, b, c) =
    let permutations = [
      (* rows *)
      (0,1,2); (0,2,1); (1,0,2); (1,2,0); (2,0,1); (2,1,0);
      (3,4,5); (3,5,4); (4,3,5); (4,5,3); (5,3,4); (5,4,3);
      (6,7,8); (6,8,7); (7,6,8); (7,8,6); (8,6,7); (8,7,6);
      (* columns *)
      (0,3,6); (0,6,3); (3,0,6); (3,6,0); (6,0,3); (6,3,0);
      (1,4,7); (1,7,4); (4,1,7); (4,7,1); (7,1,4); (7,4,1);
      (2,5,8); (2,8,5); (5,2,8); (5,8,2); (8,2,5); (8,5,2);
      (* left diagonal *)
      (0,4,8); (0,8,4); (4,0,8); (4,8,0); (8,0,4); (8,4,0); 
      (* right diagonal *)
      (2,4,6); (2,6,4); (4,2,6); (4,6,2); (6,2,4); (6,4,2); 
    ] [@@ocamlformat "disable"] in
    permutations
    |> List.find_map (fun (a', b', c') ->
           if a board.(a') && b board.(b') && c board.(c') then Some (a', b', c')
           else None)

  type t = cell array -> int option

  (* The strategy to win if 2 Os of a row, column or diagonal found *)
  let win board =
    if Array.length board != 9 then raise @@ Failure "board not length 9";
    match3 board
      ( (function E -> true | _ -> false),
        (function O (S _) -> true | _ -> false),
        function O (S _) -> true | _ -> false )
    |> Option.map (fun (empty, _, _) -> empty)

  (* The strategy to not lose if 2 Xs of a row, column or diagonal found *)
  let dont_lose board =
    if Array.length board != 9 then raise @@ Failure "board not length 9";
    match3 board
      ( (function E -> true | _ -> false),
        (function X (S _) -> true | _ -> false),
        function X (S _) -> true | _ -> false )
    |> Option.map (fun (empty, _, _) -> empty)

  (*
    The strategy to not get trapped into this particular situation by the
    opponent:

    2 Xs of a row, column or diagonal where the third is a O but after your turn
    that O will vanish
  *)
  let dont_get_trapped board =
    if Array.length board != 9 then raise @@ Failure "board not length 9";
    match3 board
      ( (function E -> true | _ -> false),
        (function O (S One) -> true | _ -> false),
        function X (S (S _)) -> true | _ -> false )
    |> Option.map (fun (empty, _, _) -> empty)

  (* The strategy to make progress by filling 2 places of a row, column or a
     diagonal, where the 3rd is either empty or will be empty soon *)
  let prepare_win board =
    if Array.length board != 9 then raise @@ Failure "board not length 9";
    match3 board
      ( (function E -> true | _ -> false),
        (function E | X One -> true | _ -> false),
        function O (S (S _)) -> true | _ -> false )
    |> Option.map (fun (empty, _, _) -> empty)

  (* The strategy when no other strategies are applicable, i.e., to just pick
     some arbitrarily interesting free cell *)
  let find_empty board =
    let rec find_empty index =
      let index = index mod 9 in
      match board.(index) with E -> index | _ -> find_empty (index + 4)
    in
    Some (find_empty 0)
end

open Game

(* diff current board with input from opponent to get X's new move *)
let detect_x_move s s' =
  let open Seq in
  zip (String.to_seq s) (String.to_seq s')
  |> find_mapi (fun i (c, c') ->
         let open Char in
         if
           equal (lowercase_ascii c') 'x' && not (equal (lowercase_ascii c) 'x')
         then Some i
         else None)

(* Start empty game *)
let game = empty

let move (str' : string) : int =
  let x_move =
    match detect_x_move game.str str' with
    | Some v -> v
    | None -> raise @@ Failure "X didn't play a move"
  in
  (* X's move *)
  move_x x_move game;
  tick_x game;
  let o_move =
    let open Strategy in
    (* Get the first strategy in order of descending priority that applies *)
    [ win; dont_lose; dont_get_trapped; prepare_win; find_empty ]
    |> List.find_map (fun strat -> strat game.board)
    (* The last strategy always applies so this never fails *)
    |> Option.get
  in
  (* O's move *)
  move_o o_move game;
  tick_o game;
  (* Save game string to be able to diff next round *)
  game.str <- string_of_board game.board;
  (* Return O's move to opponent *)
  o_move

(* C interop poggers *)
let _ = Callback.register "move" move

File: 4-tic-tac-boom/tic-tac-boom/solution_comments.ts
/*
  # The Scripting Language Tournament by InfiniteCoder01
  Challenge 4 - Tic Tac Boom
  Date: 28th August 2025

  Entry:  Typescript
  Author: Nigel Withrow <nigelwithrow78@gmail.com>

  Instructions:
  + `$ cd 4-tic-tac-boom`
  + `$ yarn`
  + `$ yarn tsc`
  + `$ sed -r -i "s/export ?\{ ?.* ?\}.*$//" solution.js` # quickjs doesn't support 'export'
  + Have system-wide Rust & Cargo installation
  + `$ cargo build --release`
  + `$ ./target/tic-tac-boom solution.js`
*/

type Cell =
  | null          // Empty
  | ["O", number] // O
  | ["X", number] // X

class Game {
  str: string
  board: Cell[]
  constructor() {
    this.str = " ".repeat(9)
    this.board = [null, null, null, null, null, null, null, null, null]
  }

  /*
    Tick lifespan of Xs and Os
  */
  tickX() {
    for (let i = 0; i < 9; i++) {
      const cell = this.board[i]
      if (cell === undefined) throw new Error("unreachable at tickX")
      if (cell === null || (cell[0] === "X" && cell[1] === 1))
        this.board[i] = null
      else if (cell[0] === "X")
        this.board[i] = ["X", cell[1] - 1]
    }
  }
  tickO() {
    for (let i = 0; i < 9; i++) {
      const cell = this.board[i]
      if (cell === undefined) throw new Error("unreachable at tickO")
      if (cell === null || (cell[0] === "O" && cell[1] === 1))
        this.board[i] = null
      else if (cell[0] === "O")
        this.board[i] = ["O", cell[1] - 1]
    }
  }

  /*
    Add moves from X and O
  */
  moveX(i: number) {
    if (this.board[i] !== null)
      throw new Error("X trying to play at occupied space")
    this.board[i] = ["X", 4]
  }
  moveO(i: number) {
    if (this.board[i] !== null)
      throw new Error("O trying to play at occupied space")
    this.board[i] = ["O", 4]
  }

  /*
    Update str to match the current game board
  */
  updateStr() {
    this.str = this.board.map((cell) => (cell === null) ? " " : cell[0]).join("")
  }
}

// Match a provided 3-cell permutation to see if any of its combinations occurs on the game board,
// returning the indexes of the matched cells if matched
function match3(
  board: Cell[],
  [a, b, c]: [(c: Cell) => boolean, (c: Cell) => boolean, (c: Cell) => boolean]
): [number, number, number] | null {
  const permutations: [number, number, number][] = [
    // rows
    [0, 1, 2], [0, 2, 1], [1, 0, 2], [1, 2, 0], [2, 0, 1], [2, 1, 0],
    [3, 4, 5], [3, 5, 4], [4, 3, 5], [4, 5, 3], [5, 3, 4], [5, 4, 3],
    [6, 7, 8], [6, 8, 7], [7, 6, 8], [7, 8, 6], [8, 6, 7], [8, 7, 6],
    // columns
    [0, 3, 6], [0, 6, 3], [3, 0, 6], [3, 6, 0], [6, 0, 3], [6, 3, 0],
    [1, 4, 7], [1, 7, 4], [4, 1, 7], [4, 7, 1], [7, 1, 4], [7, 4, 1],
    [2, 5, 8], [2, 8, 5], [5, 2, 8], [5, 8, 2], [8, 2, 5], [8, 5, 2],
    // left diagonal
    [0, 4, 8], [0, 8, 4], [4, 0, 8], [4, 8, 0], [8, 0, 4], [8, 4, 0],
    // right diagonal
    [2, 4, 6], [2, 6, 4], [4, 2, 6], [4, 6, 2], [6, 2, 4], [6, 4, 2],
  ]
  for (let i = 0; i < permutations.length; i++) {
    const permutation = permutations[i]
    if (!permutation)
      throw new Error("unreachable")
    const [a_, b_, c_] = permutation
    if (a(board[a_] as Cell) && b(board[b_] as Cell) && c(board[c_] as Cell))
      return [a_, b_, c_]
  }
  return null
}

/*
  Strategies decide what move we should play
*/
const Strategy = {
  // The strategy to win if 2 Os of a row, column or diagonal found
  win: function(board: Cell[]) {
    if (board.length !== 9) throw new Error("board not length 9")
    return match3(board, [
      (c) => c === null,
      (c) => c !== null && c[0] == "O" && c[1] > 1,
      (c) => c !== null && c[0] == "O" && c[1] > 1,
    ])?.[0];
  },

  // The strategy to not lose if 2 Xs of a row, column or diagonal found
  dontLose: function(board: Cell[]) {
    if (board.length !== 9) throw new Error("board not length 9")
    return match3(board, [
      (c) => c === null,
      (c) => c !== null && c[0] == "X" && c[1] > 1,
      (c) => c !== null && c[0] == "X" && c[1] > 1,
    ])?.[0];
  },

  /*
    The strategy to not get trapped into this particular situation by the
    opponent:

    2 Xs of a row, column or diagonal where the third is a O but after your turn
    that O will vanish
  */
  dontGetTrapped: function(board: Cell[]) {
    if (board.length !== 9) throw new Error("board not length 9")
    return match3(board, [
      (c) => c === null,
      (c) => c !== null && c[0] == "O" && c[1] === 2,
      (c) => c !== null && c[0] == "X" && c[1] > 2,
    ])?.[0];
  },

  // The strategy to make progress by filling 2 places of a row, column or a
  // diagonal, where the 3rd is either empty or will be empty soon
  prepareWin: function(board: Cell[]) {
    if (board.length !== 9) throw new Error("board not length 9")
    return match3(board, [
      (c) => c === null,
      (c) => c === null || (c[0] == "X" && c[1] === 1),
      (c) => c !== null && c[0] == "O" && c[1] > 2,
    ])?.[0];
  },

  // The strategy when no other strategies are applicable, i.e., to just pick
  // some arbitrarily interesting free cell
  findEmpty: function(board: Cell[]) {
    let index = 0
    while (true) {
      if (board[index] === null)
        return index
      index = (index + 4) % 9;
    }
  }
}

// diff current board with input from opponent to get X's new move
function detectXMove(str: string, newStr: string): number | null {
  for (let i = 0; i < 9; i++) {
    if (str[i]?.toLowerCase() !== 'x' && newStr[i]?.toLowerCase() === 'x')
      return i
  }
  return null
}

// Start new game
const game = new Game()

function move(str_: string): number {
  const xMove = detectXMove(game.str, str_)
  if (xMove === null)
    throw new Error("X didn't play a move")
  // X moves
  game.moveX(xMove)
  game.tickX()

  // Get the first strategy in order of descending priority that applies
  const strategies = [
    Strategy.win,
    Strategy.dontLose,
    Strategy.dontGetTrapped,
    Strategy.prepareWin,
    Strategy.findEmpty
  ];
  let oMove: number | undefined
  for (let i = 0; i < strategies.length; i++) {
    oMove = strategies[i]?.(game.board)
    if (oMove !== undefined) break
  }
  if (oMove === undefined)
    throw new Error("Last strategy should always apply")
  // O moves
  game.moveO(oMove)
  game.tickO()
  // Save game string to be able to diff next round
  game.updateStr()
  // Return O's move to opponent
  return oMove;
}

export { move }

